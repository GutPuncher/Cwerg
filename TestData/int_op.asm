# INT_OP

# ========================================
.fun main NORMAL [S32] = []
    .reg U32 [ru32 nu32]
    .reg U16 [ru16 nu16]
    .reg U8 [ru8 nu8]
    .reg S32 [rs32 ns32]
    .reg S16 [rs16 ns16]
    .reg S8 [rs8 ns8]

.bbl start
##########
    pusharg 10:U8
    bsr print_c_ln

    mov nu32 0
  .bbl shr_u32
    shr ru32 0x87654321:U32 nu32
    pusharg nu32
    bsr print_x_ln
    pusharg ru32
    bsr print_x_ln

    add nu32 nu32 4
    ble  nu32 64 shr_u32
##
    mov nu16 0
  .bbl shr_u16
    shr ru16 0x8765:U16 nu16
    conv ru32 nu16
    pusharg ru32
    bsr print_x_ln
    conv ru32 ru16
    pusharg ru32
    bsr print_x_ln

    add nu16 nu16 4
    ble  nu16 32 shr_u16
##
    mov nu8 0
  .bbl shr_u8
    shr ru8 0x87:U8 nu8
    conv ru32 nu8
    pusharg ru32
    bsr print_x_ln
    conv ru32 ru8
    pusharg ru32
    bsr print_x_ln

    add nu8 nu8 1
    ble  nu8 16 shr_u8
##########
    pusharg 10:U8
    bsr print_c_ln

    mov nu32 0
  .bbl shl_u32
    shl ru32 0x87654321:U32 nu32
    pusharg nu32
    bsr print_x_ln
    pusharg ru32
    bsr print_x_ln

    add nu32 nu32 4
    ble  nu32 64 shl_u32
##
    mov nu16 0
  .bbl shl_u16
    shl ru16 0x8765:U16 nu16
    conv ru32 nu16
    pusharg ru32
    bsr print_x_ln
    conv ru32 ru16
    pusharg ru32
    bsr print_x_ln

    add nu16 nu16 4
    ble  nu16 32 shl_u16
##
    mov nu8 0
  .bbl shl_u8
    shl ru8 0x87:U8 nu8
    conv ru32 nu8
    pusharg ru32
    bsr print_x_ln
    conv ru32 ru8
    pusharg ru32
    bsr print_x_ln

    add nu8 nu8 1
    ble  nu8 16 shl_u8
##########
    pusharg 10:U8
    bsr print_c_ln

    mov ns32 0
  .bbl shr_s32
    shr rs32 -2023406815:S32 ns32   # 0x87654321
    conv ru32 ns32
    pusharg ru32
    bsr print_x_ln
    conv ru32 rs32
    pusharg ru32
    bsr print_x_ln

    add ns32 ns32 4
    ble  ns32 64 shr_s32
##
    mov ns16 0
  .bbl shr_s16
    shr rs16 -30875:S16 ns16  # 0x8765
    conv ru32 ns16
    pusharg ru32
    bsr print_x_ln
    conv ru32 rs16
    pusharg ru32
    bsr print_x_ln

    add ns16 ns16 4
    ble  ns16 32 shr_s16
##
    mov ns8 0
  .bbl shr_s8
    shr rs8 -121:S8 ns8  # 0x87
    conv ru32 ns8
    pusharg ru32
    bsr print_x_ln
    conv ru32 rs8
    pusharg ru32
    bsr print_x_ln

    add ns8 ns8 1
    ble  ns8 16 shr_s8
##########
    pusharg 10:U8
    bsr print_c_ln
    
    mov ns32 0
  .bbl shr_s32_2
    shr rs32 0x12345678:S32 ns32   # 0x87654321
    conv ru32 ns32
    pusharg ru32
    bsr print_x_ln
    conv ru32 rs32
    pusharg ru32
    bsr print_x_ln

    add ns32 ns32 4
    ble  ns32 64 shr_s32_2
##
    mov ns16 0
  .bbl shr_s16_2
    shr rs16 0x1234:S16 ns16  # 0x8765
    conv ru32 ns16
    pusharg ru32
    bsr print_x_ln
    conv ru32 rs16
    pusharg ru32
    bsr print_x_ln

    add ns16 ns16 4
    ble  ns16 32 shr_s16_2
##
    mov ns8 0
  .bbl shr_s8_2
    shr rs8 0x12:S8 ns8  # 0x87
    conv ru32 ns8
    pusharg ru32
    bsr print_x_ln
    conv ru32 rs8
    pusharg ru32
    bsr print_x_ln

    add ns8 ns8 4
    ble  ns8 16 shr_s8_2
##########
    pusharg 0:S32
    ret
