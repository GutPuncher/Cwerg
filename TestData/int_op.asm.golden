

0
87654321
4
8765432
8
876543
c
87654
10
8765
14
876
18
87
1c
8
20
87654321
24
8765432
28
876543
2c
87654
30
8765
34
876
38
87
3c
8
40
87654321


0
8765
4
876
8
87
c
8
10
8765
14
876
18
87
1c
8
20
8765


0
87
2
21
4
8
6
2
8
87
a
21
c
8
e
2
10
87


0
87654321
4
76543210
8
65432100
c
54321000
10
43210000
14
32100000
18
21000000
1c
10000000
20
87654321
24
76543210
28
65432100
2c
54321000
30
43210000
34
32100000
38
21000000
3c
10000000
40
87654321


0
8765
4
7650
8
6500
c
5000
10
8765
14
7650
18
6500
1c
5000
20
8765


0
87
2
1c
4
70
6
c0
8
87
a
1c
c
70
e
c0
10
87


0
87654321
4
f8765432
8
ff876543
c
fff87654
10
ffff8765
14
fffff876
18
ffffff87
1c
fffffff8
20
87654321
24
f8765432
28
ff876543
2c
fff87654
30
ffff8765
34
fffff876
38
ffffff87
3c
fffffff8
40
87654321


0
ffff8765
4
fffff876
8
ffffff87
c
fffffff8
10
ffff8765
14
fffff876
18
ffffff87
1c
fffffff8
20
ffff8765


0
ffffff87
2
ffffffe1
4
fffffff8
6
fffffffe
8
ffffff87
a
ffffffe1
c
fffffff8
e
fffffffe
10
ffffff87


0
12345678
4
1234567
8
123456
c
12345
10
1234
14
123
18
12
1c
1
20
12345678
24
1234567
28
123456
2c
12345
30
1234
34
123
38
12
3c
1
40
12345678


0
1234
4
123
8
12
c
1
10
1234
14
123
18
12
1c
1
20
1234


0
12
2
4
4
1
6
0
8
12
a
4
c
1
e
0
10
12
